### PR TITLE
[Search] Track `discover` event and result rank

### PR DESCRIPTION
**Pull request by iamEAP**
## What / Why

A useful metric when trying to understand Backstage search usage is `Average Result Rank` (e.g. answering the question, _"When a user searches for something, what is the average position of the search result that they subsequently click on?"_)

This PR:

- Adds an optional `rank` attribute to search `Result`s, which are provided by `SearchEngine` implementations (because pagination logic is centralized there)
- In existing search result components, disables the normal `click` tracking associated with the core `<Link />` component and replaces it with a `discover` event, which includes the `rank` value.
- Updates the `app` / `create-app` so that the `rank` value is properly passed through from search results to result components.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — APPROVED** [review]
👍  from my PoV

Is this something you'd like to get shipped in 1.3?

**Comment by iamEAP** [comment]
> 👍 from my PoV
> 
> Is this something you'd like to get shipped in 1.3?

@Rugvip _Ideally_ yes!  But priority is on #11849 (which will conflict somewhat with this, and for which I will fix conflicts after merge)

File: .changeset/search-rankly-my-dear.md
Comment (emmaindal) [inline .changeset/search-rankly-my-dear.md:7]: So I guess this mean that click events doesn't get captured anymore? Because of that change I wonder if this should either be changed to an optional replacement OR we should mark this as breaking? As this could be used for analysis already by adopters?

File: docs/plugins/analytics.md
Comment (emmaindal) [inline docs/plugins/analytics.md:60]: I'm a bit unsure about the "provided by" column and more now when its mixed with core, plugins and something else. Is this meant to tell me what part of Backstage I can use this actions? If so can we clarify what "by convention" means here?

**Review by emmaindal — APPROVED** [review]
Looks good to me, just smaller comments but will approve to unblock!

File: docs/plugins/analytics.md
Comment (iamEAP) [reply docs/plugins/analytics.md:60]: Hmm...  Yeah, good point.  I guess the goal of the column is to inform _where_ the event is being fired from?  So that, if you saw the events start appearing in your analytics platform, you'd know where to look for what it meant.  ...Or if you are interested in being able to measure certain things, you'd know what you'd need to install to start getting the data.

...And from that perspective, `(By Convention)` is not super helpful.  And `Search Plugin` is kind of a misnomer too, because I think the event is actually fired within `@backstage/plugin-search-react`.  We could link directly to places in code where events are fired, but that'd be difficult to maintain.

Maybe we just get rid of the column entirely?  I'm not sure it's serving its purpose right now.

**Pull request merged**
